src/librustdoc/html/format.rs

```diff
@@ -456,15 +456,26 @@ impl clean::GenericArgs {
     }
 }
 
-crate fn href(did: DefId, cx: &Context<'_>) -> Option<(String, ItemType, Vec<String>)> {
+// Possible errors when computing href link source for a `DefId`
+crate enum HrefError {
+    // `DefId` is in an unknown location. This seems to happen when building without dependencies
+    // but a trait from a dependency is still visible
+    UnknownLocation,
+    // Unavailable because private
+    Unavailable,
+    // Not in external cache, href link should be in same page
+    NotInExternalCache,
+}
+
+crate fn href(did: DefId, cx: &Context<'_>) -> Result<(String, ItemType, Vec<String>), HrefError> {
     let cache = &cx.cache();
     let relative_to = &cx.current;
     fn to_module_fqp(shortty: ItemType, fqp: &[String]) -> &[String] {
         if shortty == ItemType::Module { &fqp[..] } else { &fqp[..fqp.len() - 1] }
     }
 
     if !did.is_local() && !cache.access_levels.is_public(did) && !cache.document_private {
```

Comment:
So, `!is_local` looks correct to me, but for a non-trivial reason - the `strip_private` pass should get rid of private local items before they ever get here. Can you change this to an assert instead?
```suggestion
    if did.is_local() {
        // Private items should have already been stripped.
        assert!(cache.access_levels.is_public(did) || cache.document_private);
    }
    if !cache.access_levels.is_public(did) && !cache.document_private {
```
(I know this is unrelated to your change - I can make a separate PR for this if you like.)

Reply:
adding this assert made the following test fail:
```
failures:
    [rustdoc] rustdoc/all.rs
    [rustdoc] rustdoc/async-fn.rs
    [rustdoc] rustdoc/blanket-reexport-item.rs
    [rustdoc] rustdoc/impl-trait-alias.rs
    [rustdoc] rustdoc/inline_cross/renamed-via-module.rs
    [rustdoc] rustdoc/inline_local/glob-private-no-defaults.rs
    [rustdoc] rustdoc/inline_local/glob-private.rs
    [rustdoc] rustdoc/inline_local/issue-28537.rs
    [rustdoc] rustdoc/inline_local/issue-32343.rs
    [rustdoc] rustdoc/inline_local/trait-vis.rs
    [rustdoc] rustdoc/intra-doc/cross-crate/hidden.rs
    [rustdoc] rustdoc/issue-34473.rs
    [rustdoc] rustdoc/issue-35488.rs
    [rustdoc] rustdoc/issue-46767.rs
    [rustdoc] rustdoc/namespaces.rs
    [rustdoc] rustdoc/redirect-map.rs
    [rustdoc] rustdoc/redirect-rename.rs
    [rustdoc] rustdoc/redirect.rs
    [rustdoc] rustdoc/search-index.rs
    [rustdoc] rustdoc/struct-arg-pattern.rs
    [rustdoc] rustdoc/synthetic_auto/complex.rs
    [rustdoc] rustdoc/synthetic_auto/project.rs
```

Reply:
ahhhh of course it did :( I'll open an issue

Reply:
Oh, I was just confused: the proper assert is
```
        assert!(cache.access_levels.is_public(did) || cache.document_private || cache.paths.get(&did).is_none(), "{:?}", did);
```

Reply:
well, there was a bug after all: https://github.com/rust-lang/rust/issues/87048